File: frigate-camera-app/app/src/main/java/com/example/frigatecamerapp/MainActivity.java
package com.example.frigatecamerapp;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.view.PreviewView;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.pm.PackageManager;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.text.format.Formatter;
import android.util.Log;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.pedro.rtspserver.RtspServerCamera2;
import com.pedro.rtspserver.util.ConnectCheckerRtsp;

/**
 * Main activity for the Frigate Camera App.
 * This activity handles camera permissions, displays the camera preview,
 * and manages the RTSP server for streaming video.
 */
public class MainActivity extends AppCompatActivity implements ConnectCheckerRtsp {

    private static final String TAG = "MainActivity";
    private final String[] REQUIRED_PERMISSIONS = new String[]{Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO};
    private static final int RTSP_PORT = 1935; // Standard RTSP port, can be changed.

    private RtspServerCamera2 rtspServerCamera2;
    private Button startStopButton;
    private TextView infoText;
    private PreviewView previewView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Keep the screen on while the app is active
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        setContentView(R.layout.activity_main);

        previewView = findViewById(R.id.previewView);
        startStopButton = findViewById(R.id.start_stop_button);
        infoText = findViewById(R.id.info_text);

        // Check for camera and audio permissions. If granted, set up the server.
        // Otherwise, request permissions.
        if (allPermissionsGranted()) {
            setupRtspServer();
        } else {
            requestPermissionsLauncher.launch(REQUIRED_PERMISSIONS);
        }
    }

    /**
     * Initializes the RTSP server and sets up the start/stop button.
     */
    private void setupRtspServer() {
        rtspServerCamera2 = new RtspServerCamera2(previewView, this, RTSP_PORT);
        startStopButton.setOnClickListener(v -> {
            if (!rtspServerCamera2.isStreaming()) {
                // Prepare video and audio encoders
                if (rtspServerCamera2.prepareAudio() && rtspServerCamera2.prepareVideo()) {
                    // Start the stream
                    startStopButton.setText("Stop Stream");
                    rtspServerCamera2.startStream();
                    updateUrl(); // Update the URL text view
                } else {
                    Toast.makeText(this, "Error preparing stream. Check logs for details.", Toast.LENGTH_SHORT).show();
                }
            } else {
                // Stop the stream
                startStopButton.setText("Start Stream");
                rtspServerCamera2.stopStream();
                infoText.setText("Stream stopped");
            }
        });
        updateUrl(); // Set initial URL text
    }

    private boolean allPermissionsGranted() {
        for (String permission : REQUIRED_PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    /**
     * Modern way to handle permission requests.
     */
    private final ActivityResultLauncher<String[]> requestPermissionsLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), result -> {
                if (Boolean.TRUE.equals(result.get(Manifest.permission.CAMERA)) && Boolean.TRUE.equals(result.get(Manifest.permission.RECORD_AUDIO))) {
                    // Permissions granted, set up the server
                    setupRtspServer();
                } else {
                    // Permissions denied, show a message and close the app
                    Toast.makeText(this, "Permissions not granted. App will close.", Toast.LENGTH_LONG).show();
                    finish();
                }
            });

    /**
     * Gets the local IP address of the device.
     * @return The formatted IP address or "0.0.0.0" if not available.
     */
    private String getIpAddress() {
        try {
            WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(WIFI_SERVICE);
            return Formatter.formatIpAddress(wifiManager.getConnectionInfo().getIpAddress());
        } catch (Exception e) {
            Log.e(TAG, "Failed to get IP address", e);
            return "0.0.0.0";
        }
    }

    /**
     * Updates the TextView to display the current RTSP stream URL.
     */
    private void updateUrl() {
        String ip = getIpAddress();
        String url = "rtsp://" + ip + ":" + RTSP_PORT + "/";
        infoText.setText(url);
    }

    // ConnectCheckerRtsp callbacks
    // These are called from the RTSP server library thread, so we use runOnUiThread to update the UI.

    @Override
    public void onConnectionSuccessRtsp() {
        runOnUiThread(() -> Toast.makeText(MainActivity.this, "Stream started successfully", Toast.LENGTH_SHORT).show());
    }

    @Override
    public void onConnectionFailedRtsp(String reason) {
        runOnUiThread(() -> {
            Toast.makeText(MainActivity.this, "Connection failed: " + reason, Toast.LENGTH_SHORT).show();
            rtspServerCamera2.stopStream();
            startStopButton.setText("Start Stream");
            infoText.setText("Stream stopped");
        });
    }

    @Override
    public void onNewBitrateRtsp(long bitrate) {
        // This callback can be used to monitor the stream bitrate in real-time.
        // Not used in this basic implementation.
    }

    @Override
    public void onDisconnectRtsp() {
        runOnUiThread(() -> {
            Toast.makeText(MainActivity.this, "Stream disconnected", Toast.LENGTH_SHORT).show();
            startStopButton.setText("Start Stream");
            infoText.setText("Stream stopped");
        });
    }

    @Override
    public void onAuthErrorRtsp() {
        runOnUiThread(() -> Toast.makeText(MainActivity.this, "Auth error", Toast.LENGTH_SHORT).show());
    }

    @Override
    public void onAuthSuccessRtsp() {
        runOnUiThread(() -> Toast.makeText(MainActivity.this, "Auth success", Toast.LENGTH_SHORT).show());
    }
}

File: frigate-camera-app-v11/app/src/main/java/com/example/frigatecamerappv11/MainActivity.kt
package com.example.frigatecamerappv11

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.SurfaceTexture
import android.os.Build
import android.os.Bundle
import android.view.TextureView
import android.view.WindowManager
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.pedro.common.ConnectChecker
import com.pedro.encoder.input.video.CameraHelper
import com.pedro.encoder.input.video.CameraOpenException
import com.pedro.library.view.AutoFitTextureView
import com.pedro.rtspserver.RtspServerCamera1
import com.pedro.rtspserver.server.ClientListener
import com.pedro.rtspserver.server.ServerClient

class MainActivity : AppCompatActivity(), ConnectChecker, ClientListener, TextureView.SurfaceTextureListener {

    private lateinit var rtspServerCamera1: RtspServerCamera1
    private lateinit var surfaceView: AutoFitTextureView
    private lateinit var bStartStop: Button
    private lateinit var bSwitchCamera: Button
    private lateinit var tvUrl: TextView

    private val permissions = arrayOf(Manifest.permission.RECORD_AUDIO, Manifest.permission.CAMERA)
    private val PERMISSIONS_REQUEST_CODE = 1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        setContentView(R.layout.activity_main)

        surfaceView = findViewById(R.id.surfaceView)
        bStartStop = findViewById(R.id.b_start_stop)
        bSwitchCamera = findViewById(R.id.switch_camera)
        tvUrl = findViewById(R.id.tv_url)

        if (!hasPermissions(this, *permissions)) {
            ActivityCompat.requestPermissions(this, permissions, PERMISSIONS_REQUEST_CODE)
        } else {
            setupRtsp()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSIONS_REQUEST_CODE) {
            if (grantResults.all { it == PackageManager.PERMISSION_GRANTED }) {
                setupRtsp()
            } else {
                toast("Permissions not granted")
                finish()
            }
        }
    }

    private fun setupRtsp() {
        rtspServerCamera1 = RtspServerCamera1(surfaceView, this, 1935)
        rtspServerCamera1.streamClient.setClientListener(this)
        surfaceView.surfaceTextureListener = this

        bStartStop.setOnClickListener {
            if (!rtspServerCamera1.isStreaming) {
                if (rtspServerCamera1.prepareAudio() && rtspServerCamera1.prepareVideo()) {
                    bStartStop.text = "Stop"
                    rtspServerCamera1.startStream()
                    tvUrl.text = rtspServerCamera1.streamClient.getEndPointConnection()
                } else {
                    toast("Error preparing stream, This device cant do it")
                }
            } else {
                bStartStop.text = "Start"
                rtspServerCamera1.stopStream()
                tvUrl.text = ""
            }
        }

        bSwitchCamera.setOnClickListener {
            try {
                rtspServerCamera1.switchCamera()
            } catch (e: CameraOpenException) {
                toast(e.message ?: "Unknown camera error")
            }
        }
    }

    private fun hasPermissions(activity: AppCompatActivity, vararg permissions: String): Boolean = permissions.all {
        ActivityCompat.checkSelfPermission(activity, it) == PackageManager.PERMISSION_GRANTED
    }

    private fun toast(text: String) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show()
    }

    private fun adaptPreview() {
        val isPortrait = CameraHelper.isPortrait(this)
        val w = if (isPortrait) rtspServerCamera1.streamHeight else rtspServerCamera1.streamWidth
        val h = if (isPortrait) rtspServerCamera1.streamWidth else rtspServerCamera1.streamHeight
        surfaceView.setAspectRatio(w, h)
    }

    override fun onConnectionSuccess() {
        toast("Connection success")
    }

    override fun onConnectionFailed(reason: String) {
        toast("Connection failed: $reason")
        rtspServerCamera1.stopStream()
        bStartStop.text = "Start"
    }

    override fun onConnectionStarted(url: String) {
    }

    override fun onDisconnect() {
        toast("Disconnected")
    }

    override fun onAuthError() {
        toast("Auth error")
    }

    override fun onAuthSuccess() {
        toast("Auth success")
    }

    override fun onClientConnected(client: ServerClient) {
        toast("Client connected: ${client.address}")
    }

    override fun onClientDisconnected(client: ServerClient) {
        toast("Client disconnected: ${client.address}")
    }

    override fun onClientNewBitrate(bitrate: Long, client: ServerClient) {}

    override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
        if (!rtspServerCamera1.isOnPreview) {
            rtspServerCamera1.startPreview()
            adaptPreview()
        }
    }

    override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
    }

    override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
        if (rtspServerCamera1.isStreaming) {
            rtspServerCamera1.stopStream()
            bStartStop.text = "Start"
        }
        if (rtspServerCamera1.isOnPreview) {
            rtspServerCamera1.stopPreview()
        }
        return true
    }

    override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {
    }
}
